models/stock_picking_batch.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
+
+
+class StockPickingBatch(models.Model):
+    _inherit = 'stock.picking.batch'
+
+    def _check_user_id(self, user_id):
+        if user_id is None:
+            user_id = self.env.user.id
+
+        if not user_id:
+            raise ValidationError(_("Cannot determine the user"))
+
+        return user_id
+
+
+    @api.multi
+    def get_single_batch(self, user_id=None):
+        """
+        Search for a picking batch in progress for the specified user.
+        If no user is specified, the current user is considered.
+        If no batch is found, but pickings exist, create a new batch.
+
+        If a batch is determined, return it, otherwise return None.
+
+        Raise a ValidationError in case it cannot perform a search
+        or if multiple batches are found for the specified user.
+        """
+        PickingBatch = self.env['stock.picking.batch']
+
+        user_id = self._check_user_id(user_id)
+        batches = PickingBatch.search([('user_id', '=', user_id),
+                                       ('state', '=', 'in_progress')])
+        batch = None
+
+        if not batches:
+            batch = self._create_batch(user_id)
+        else:
+            if len(batches) > 1:
+                raise ValidationError(
+                    _("Expected single picking batch, found %d batches")
+                    % len(batches))
+
+            batch = batches[0]
+
+        return batch
+
+    def _get_single_batch_info(self, batch, allowed_picking_states):
+        filtered_pickings = batch.picking_ids.filtered(
+            lambda x: x.state in allowed_picking_states)
+
+        return {'id': self.id,
+                'picking_ids': filtered_pickings.get_info()}
+
+    def get_info(self, allowed_picking_states=None):
+        """
+        Return list of dictionaries containing information about
+        all batches.
+        """
+        if allowed_picking_states is None:
+            allowed_picking_states = ['assigned']
+
+        return [self._get_single_batch_info(batch, allowed_picking_states)
+                for batch in self]
+
+    @api.multi
+    def create_batch(self, picking_priorities, user_id=None):
+        """
+        Creeate and return a batch for the specified user if pickings
+        exist. Return None otherwise. Pickings are filtered based on
+        the specified picking priorities.
+
+        Raise a ValidationError in case the user ID is not specified.
+
+        If the user already has batches assigned, a ValidationError
+        is raised in case of pickings that need to be completed,
+        otherwise such batches will be marked as done.
+        """
+        user_id = self._check_user_id(user_id)
+        self._check_batches(user_id)
+
+        return self._create_batch(user_id, picking_priorities)
+
+    def _create_batch(self, user_id, picking_priorities=None):
```

Comment:
Add docstring?

Reply:
A docstring for this private method would have to include everything you have in the public `create_batch` one, which you want to have because it's public. But in that case you end up having two sources of truth to maintain, so... It's a bit of a trade off. No strong opinion though.